crates/matrix-sdk-ui/src/timeline/event_handler.rs

```diff
@@ -1150,13 +1155,27 @@ impl<'a, 'o> TimelineEventHandler<'a, 'o> {
                     None => self.meta.new_timeline_item(item),
                 };
 
-                // Keep push semantics, if we're inserting at the front or the back.
-                if insert_idx == self.items.len() {
-                    self.items.push_back(new_item);
-                } else if insert_idx == 0 {
+                // We are about to insert the `new_item`, great! Though, we try to keep a
+                // precise insertion semantics here:
+                //
+                // * _push front_:
+                //   * if `must_push_front`, i.e. if the position is a `TimelinePosition::Start`
+                //   * if the index is 0 _but_ the items aren't empty
+                //
+                // * _push back_:
+                //   * if the index represents the end of items
+                //
+                // * _insert at_:
+                //   * for any other index value
+                if must_push_front || (insert_index == 0 && self.items.is_empty().not()) {
```

Comment:
I found the comment confusing, but that's because the code is a bit confusing too: we don't need `must_push_front`, since the insert_index will be correct in that case. I think you can actually keep the previous code (that preferred push back over push front, see comment below) and it would just work fine. 

Once that's confirmed working, can you keep the previous comment, maybe tweak it to explain why it's not a coincidence that we start with `push_back`?

Reply:
Rephrased.